malsim/mal_simulator.py

```diff
@@ -33,66 +34,88 @@ class MalSimAgentState:
     # Attackers get positive rewards, defenders negative
     reward: int = 0
 
+    # Contains the steps performed successfully in the last step
+    step_performed_nodes: set[AttackGraphNode] = field(default_factory=set)
+
     # Contains possible actions for the agent in the next step
-    action_surface: list[AttackGraphNode] = field(default_factory=list)
+    action_surface: set[AttackGraphNode] = field(default_factory=set)
+
+    # Contains possible actions that became available in the last step
+    step_action_surface_additions: set[AttackGraphNode] = (
+        field(default_factory = set))
+
+    # Contains previously possible actions that became unavailable in the last
+    # step
+    step_action_surface_removals: set[AttackGraphNode] = (
+        field(default_factory = set))
 
-    # Fields that tell if agent is 'dead' / disabled
+    # Contains nodes that defender actions made unviable in the last step
+    step_unviable_nodes: set[AttackGraphNode] = field(default_factory=set)
```

Comment:
This is the same as `step_action_surface_removals` minus the enabled defense steps, right? It is a global view. Do you think it belongs to the attackers' state?